### ISSUE TITLE
Authenticated NuGet feeds: 401 reported as "package not found", and package search returns empty against every Azure DevOps feed

### ISSUE THREAD

**Issue by rjmurillo** [richlander/dotnet-inspect#3417]
## Ask first

I would like to add support for **authenticated NuGet feeds** (Azure Artifacts / GitHub Packages), so that `dotnet inspect package` and `dotnet inspect package search` work against a private feed without a prior `dotnet package download` or `dotnet restore` to warm the global packages folder.

Before writing any code I want scope agreement, because the work splits into **three bug fixes and one feature**, and the bugs are independently valuable. Please tell me which parts you want, and whether you want them as one PR or several.

All four items below are reproduced against `0.16.0+44f2832` on Windows using **public, credential-free feeds**, so you can verify each one without an internal account.

---

## Bug 1: an HTTP 401 from a source is reported as "package not found"

`https://pkgs.dev.azure.com/dnceng/_packaging/DoesNotExistPubliclyXYZ/nuget/v3/index.json` returns **HTTP 401 Unauthorized** (Azure DevOps answers 401 rather than 404 for a feed the caller cannot see).

```console
$ dotnet-inspect package NodaTime.Testing@3.1.9 --source https://pkgs.dev.azure.com/dnceng/_packaging/DoesNotExistPubliclyXYZ/nuget/v3/index.json
Error: Package 'nodatime.testing' not found.
(exit 1)
```

The package exists; the *feed* was not readable. The message names the wrong subject, and there is no indication that authentication was the problem, which source failed, or at which phase.

Path: `HttpRetryHelper` classifies 401/403 as non-retryable and returns `null` (`src/DotnetInspector.Packages/HttpRetryHelper.cs:118-132`); `PackageExtractor` then treats the null as absence and prints "not found" (`src/DotnetInspector.Packages/PackageExtractor.cs:224-228`, `:330-347`). Source identity and failure phase are lost on the way out because the current signatures return `string?` / `List<string>?` / `bool`.

This is the defect that makes an authenticated feed look like a typo, and it is worth fixing on its own even if you want none of the rest.

Note: `--source` itself behaves correctly. It **does** replace the defaults; the command above does not silently fall back to nuget.org. A separate, much milder observation is that a package already present in `~/.nuget/packages` is returned with `Source: NuGet` and exit 0 even when the requested `--source` is unreachable, so a cache hit hides an unusable source.

## Bug 2: `package search` returns zero results against any Azure DevOps feed

Azure DevOps serializes `totalHits` as a **JSON string**; nuget.org serializes it as a **number**. Both are otherwise valid NuGet V3 search responses.

```
nuget.org     "totalHits":1032     (number, first property)
Azure DevOps  "totalHits":"0"      (string, last property, and "0" even when data is returned)
```

Public reproduction, no credentials:

```console
$ curl -s "https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/query2/?q=System.Text.Json&take=5"
{"data":[ ... 1 real result ... ],"totalHits":"0"}
```

Note `data` is populated while `totalHits` reports `"0"`.

`NuGetFetch.SearchResponse.TotalHits` is typed `int`. Deserializing both bodies through NuGetFetch's own source-generated `NuGetJsonContext`:

```
nuget.org      OK     TotalHits=1032  Data.Count=1
Azure DevOps   THROW  JsonException: The JSON value could not be converted to
                      System.Int32. Path: $.totalHits
```

`NuGetApi.GetSearchResponseAsync` catches the `JsonException` and returns `null`; `SearchService.SearchAsync` then returns `response?.Data ?? Array.Empty<SearchResult>()`. Real results become zero, with no error surfaced. Confirmed end to end against a populated Azure DevOps feed: the same URL with the same credential returns results through a raw `HttpClient` and 0 results through `SearchService`.

This is not an authentication bug. It breaks **every** Azure DevOps feed, public ones included, and `"totalHits":"0"` alongside a populated `data` array means the field is not trustworthy even when it does parse.

## Bug 3: `package search` cannot target a source at all

```console
$ dotnet-inspect package search --help
Options:
  --take <take>            Maximum number of results (default: 20)
  --preview, --prerelease  Include prerelease versions
  --json                   Output as JSON
  --compact                Minified JSON (use with --json)
  --verbose                Show progress messages on stderr
  -n, --head <n>           Limit to first N lines (like head -n)
```

There is no `--source`, `--add-source`, or `--nugetconfig`, so search can only ever reach nuget.org, while its sibling `package` command accepts all three. Even with Bug 2 fixed, search cannot be pointed at a private feed.

## Bug 4: `package search` exits 0 when it finds nothing

`src/dotnet-inspect/Commands/PackageSearchCommand.cs:40-43`:

```csharp
if (results.Count == 0)
{
    Console.Error.WriteLine($"No packages found for \"{options.Query}\".");
    return 0;
}
```

Composed with Bug 2, a hard `JsonException` renders as a successful empty search: exit 0, "No packages found". That is a failure shaped like success, which `AGENTS.md` ("Keep failure visible") rules out. Whether a genuine zero-result search should also be non-zero is your call; the swallowed-exception case clearly should be.

---

## Feature: acquire credentials for authenticated feeds

PR #192 added static credentials from `nuget.config`. That does not cover Azure Artifacts, where tokens are short-lived and issued by the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider), which `dotnet restore` invokes but `dotnet-inspect` does not.

Proposed shape, kept behind the existing progressive-disclosure and threat-model rules:

- A `INuGetCredentialResolver` seam in `DotnetInspector.Packages`, resolving in order: existing `nuget.config` credentials, then `NUGET_CREDENTIALPROVIDER_*` / env tokens, then the installed credential provider plugin.
- Provider invocation is **explicit or capability-gated**, never automatic on an arbitrary 401, and only for a source the user named.
- Tokens held in memory for the process lifetime, attached per-source, never logged, never written to output. (The provider maintains its own on-disk caches; those are the same ones `dotnet restore` already populates, and I do not propose changing them.)
- Retry once with credentials on a 401 from a user-specified source, then fail loudly with source, phase, and status.

Open design questions I would want your opinion on:

1. Should provider launch be gated on the existing network capability, or a new one?
2. `PackageExtractor` currently trusts feed-supplied `@id` values for `PackageBaseAddress` and `SearchQueryService` (`:527-580`). Sending a credential to a feed-controlled URL needs a gate. Is a same-origin-as-service-index check acceptable, or do you want something stricter?
3. With multiple `--source` values, if one 401s and another succeeds, is that success, partial success, or failure?

## Proposed split

| PR | Contents | Depends on |
|---|---|---|
| 1 | Typed failure model carrying source, phase, status, failure kind; Bug 1 error messages | none |
| 2 | Bug 4 exit code | none |
| 3 | Bug 2 tolerant search-response parsing + Bug 3 `--source` on search | 1 |
| 4 | Credential resolver + provider integration | 1 |

PRs 1, 2, and 3 contain no authentication code and are useful on their own. I originally hoped Bug 1 was severable from the typed failure model; it is not, because the current signatures cannot carry the source or the phase.

I have a branch and a build set up and am happy to do all four, a subset, or none if you would rather own this. **Which parts do you want, and how would you like them split?**

## Environment

- `dotnet-inspect` `0.16.0+44f2832`
- .NET SDK `11.0.100-preview.6.26359.118`, Windows 11
- Search parsing checked against `NuGetFetch`'s own source-generated `NuGetJsonContext`, both feeds, same code path

**Comment by richlander** [comment]
Bug 1 still reproduces on merged `main` (`fdf24f5f7`), now against live Azure DevOps infrastructure rather than the `dnceng` example in the description. Recording it because the credential feature merging (#3428) changes how often this is hit, and not in the direction I expected.

## Reproduction

A real private Azure Artifacts feed, with a PAT that has since expired:

```console
$ dotnet-inspect package Markout --source https://pkgs.dev.azure.com/<org>/<project>/_packaging/<feed>/nuget/v3/index.json
Error: Package 'markout' not found.
```

Same result with a valid `nuget.config` supplying `Username` + `ClearTextPassword`, because the credential itself is the thing that is dead.

What the feed actually answers:

```console
$ curl -s -D- -o /dev/null -H "Authorization: Basic $(printf 'pat:%s' "$PAT" | base64 -w0)" "$FEED"
HTTP/2 401
www-authenticate: Basic realm="https://pkgsprodcin1.pkgs.visualstudio.com/"
```

So: an authenticated 401 with a Basic challenge, reported as a missing package.

## Path confirmed at `fdf24f5f7`

The description's trace still holds, at slightly different line numbers after the #3423 rehome:

- `HttpRetryHelper.cs:129-133` treats 401 as non-retryable and returns `new HttpRetryResult(null, statusCode)`. The status **is** captured here.
- `PackageExtractor.cs:222-228` calls `GetLatestVersionAsync`, whose return type is `string?`. The status cannot survive that signature, so `null` arrives and becomes `Package '<name>' not found.`

The information exists at the point of failure and is discarded one frame later, which is why this needs the typed failure model in PR 1 of the split rather than a message change.

## Why this is more urgent after #3428

Before credential support, reaching this message meant you had no credentials configured, and "not found" was merely unhelpful.

After #3428, users will configure credentials, and **an expired credential is the normal end state of every PAT**. Azure DevOps is actively shortening that lifetime — `docs/design/nuget-authentication.md` already links the [global PAT retirement](https://devblogs.microsoft.com/devops/retirement-of-global-personal-access-tokens-in-azure-devops/) (December 1, 2026, with a push toward short-lived Entra-backed auth). Short-lived credentials mean expiry stops being an edge case and becomes routine.

That makes `Package 'markout' not found.` the message users will most often see from the feature that just shipped, at the exact moment the correct action is "refresh your credential." It is indistinguishable from a typo in the package name.

Found while running an end-to-end check of #3428 against a live private feed after merge.

**Comment by richlander** [comment]
Status check on the other three bugs, since bug 1 (#3563) is the only one I have been tracking actively. All three are **already fixed on `main`**; I verified each rather than going by the commit log.

**Bug 2 — search returns zero against Azure DevOps: fixed.** `TotalHits` was removed from the `SearchResponse` model entirely (`src/NuGetFetch/NuGetRecords.cs:30`), so the string/number discrepancy can no longer reject the document, with a regression test at `src/NuGetFetch.Tests/NuGetApiTests.cs:177`. The upstream behaviour is unchanged — the public feed still returns a populated `data` array beside `"totalHits":"0"`:

```console
$ curl -s ".../nuget/v3/query2/?q=System.Text.Json&take=2" | tail -c 120
..."index":"PackageIndex","totalHits":"0"}
```

End to end against that same public Azure DevOps feed, which is the case that used to return nothing:

```console
$ dotnet-inspect package search System.Text.Json --source https://pkgs.dev.azure.com/dnceng/.../nuget/v3/index.json --take 3
| System.Text.Json | 10.0.10 | 0 | Provides high-performance and low-allocating types that ser... |
```

**Bug 3 — search cannot target a source: fixed.** `package search --help` now lists `--source`, `--add-source` and `--nugetconfig`, matching its sibling `package` command.

**Bug 4 — exit 0 on empty: fixed.** A source that cannot be searched now produces a warning and a non-zero exit, so an incomplete search no longer reads like a complete one:

```console
$ dotnet-inspect package search markout --source <private AzDO feed>
Error: No configured NuGet source could be searched.
exit=1
```

So bug 1 is the last one open, and #3563 is its fix. I have updated that PR to close this issue on merge.

One thing that came out of checking bug 4: the message above blames "service index unavailable, or advertises no SearchQueryService", but the feed actually returned **401**. That is bug 1's misdiagnosis on the search path, and `NuGetSourceCompat.cs:278-286` carries a comment saying it is waiting on precisely the typed-failure mechanism #3563 introduces. Filed separately as #3572 rather than widening #3563 mid-review.
